File: versions/Python/micromoth.py
import random
from math import cos,sin,pi
r2=0.70710678118 
class QuantumCircuit:
  def __init__(self,n,m=0):
    self.num_qubits=n
    self.num_clbits=m
    self.name = ''
    self.data=[]
  def __add__(self,self2):
    self3=QuantumCircuit(max(self.num_qubits,self2.num_qubits),max(self.num_clbits,self2.num_clbits))
    self3.data=self.data+self2.data
    self3.name = self.name
    return self3
  def initialize(self,k):
    self.data[:] = [] 
    self.data.append(('init',[e for e in k])) 
  def x(self,q):
    self.data.append(('x',q))
  def rx(self,theta,q):
    self.data.append(('rx',theta,q))
  def rz(self,theta,q):
    self.data.append(('rz',theta,q))
  def h(self,q):
    self.data.append(('h',q))
  def cx(self,s,t):
    self.data.append(('cx',s,t))
  def crx(self,theta,s,t):
    self.data.append(('crx',theta,s,t))
  def swap(self,s,t):
    self.data.append(('swap',s,t))
  def measure(self,q,b):
    assert b<self.num_clbits, 'Index for output bit out of range.'
    assert q<self.num_qubits, 'Index for qubit out of range.'
    self.data.append(('m',q,b))
  def measure_all(self):
      if self.num_clbits==0:
        self.num_clbits = self.num_qubits
      for q in range(self.num_qubits):
        self.measure(q,q)
  def ry(self,theta,q):
    self.rx(pi/2,q)
    self.rz(theta,q)
    self.rx(-pi/2,q)
  def z(self,q):
    self.rz(pi,q)
  def t(self,q):
    self.rz(pi/4,q)
  def y(self,q):
    self.rz(pi,q)
    self.x(q)
  def qasm(self):
    qasm = 'OPENQASM 2.0;\ninclude "qelib1.inc";\n\n'
    qasm += 'qreg q['+str(self.num_qubits)+'];\n'
    if self.num_clbits:
      qasm += 'creg c['+str(self.num_clbits)+'];\n'
    qasm += '\n'
    for gate in self.data:
      if gate[0] in ['x', 'h']:
        qasm += gate[0]+' q['+str(gate[1])+'];\n'
      elif gate[0] in ['rx', 'rz']:
        qasm += gate[0]+'('+str(gate[1])+')  q['+str(gate[1])+'];\n'
      elif gate[0]=='m':
        qasm += 'measure q['+str(gate[1])+'] -> c['+str(gate[2])+'];\n'
      elif gate[0] in ['cx', 'swap']:
        qasm += gate[0]+' q['+str(gate[1])+'], q['+str(gate[2])+'];\n'
      elif gate[0]=='crx':
        qasm += 'cz q['+str(gate[2])+'], q['+str(gate[3])+'];\n'
        qasm += 'rx('+str(-gate[0]/2)+')  q['+str(gate[3])+'];\n'
        qasm += 'cz q['+str(gate[2])+'], q['+str(gate[3])+'];\n'
        qasm += 'rx(-'+str(gate[0]/2)+')  q['+str(gate[3])+'];\n'
    return qasm
def simulate(qc,shots=1024,get='counts',noise_model=[]):
  def superpose(x,y):
    return [r2*(x[j]+y[j])for j in range(2)],[r2*(x[j]-y[j])for j in range(2)]
  def turn(x,y,theta):
    theta = float(theta)
    return [x[0]*cos(theta/2)+y[1]*sin(theta/2),x[1]*cos(theta/2)-y[0]*sin(theta/2)],[y[0]*cos(theta/2)+x[1]*sin(theta/2),y[1]*cos(theta/2)-x[0]*sin(theta/2)]
  def phaseturn(x,y,theta):
     theta = float(theta)
     return [[x[0]*cos(theta/2) - x[1]*sin(-theta/2),x[1]*cos(theta/2) + x[0]*sin(-theta/2)],[y[0]*cos(theta/2) - y[1]*sin(+theta/2),y[1]*cos(theta/2) + y[0]*sin(+theta/2)]]   
  k = [[0,0] for _ in range(2**qc.num_qubits)] 
  k[0] = [1.0,0.0] 
  if noise_model:
    if type(noise_model)==float:
       noise_model = [noise_model]*qc.num_qubits
  outputnum_clbitsap = {}
  for gate in qc.data:
    if gate[0]=='init': 
      if type(gate[1][0])==list:
        k = [e for e in gate[1]]
      else: 
        k = [[e,0] for e in gate[1]]
    elif gate[0]=='m': 
      outputnum_clbitsap[gate[2]] = gate[1]
    elif gate[0] in ['x','h','rx','rz']: 
      j = gate[-1] 
      for i0 in range(2**j):
        for i1 in range(2**(qc.num_qubits-j-1)):
          b0=i0+2**(j+1)*i1 
          b1=b0+2**j 
          if gate[0]=='x': 
            k[b0],k[b1]=k[b1],k[b0]
          elif gate[0]=='h': 
            k[b0],k[b1]=superpose(k[b0],k[b1])
          elif gate[0]=='rx': 
            theta = gate[1]
            k[b0],k[b1]=turn(k[b0],k[b1],theta)
          elif gate[0]=='rz': 
            theta = gate[1]
            k[b0],k[b1]=phaseturn(k[b0],k[b1],theta) 
    elif gate[0] in ['cx','crx']: 
      if gate[0]=='cx': 
        [s,t] = gate[1:]
      else:
        theta = gate[1]
        [s,t] = gate[2:]
      [l,h] = sorted([s,t])
      for i0 in range(2**l):
        for i1 in range(2**(h-l-1)):
          for i2 in range(2**(qc.num_qubits-h-1)):
            b00=i0+2**(l+1)*i1+2**(h+1)*i2 
            b01=b00+2**t 
            b10=b00+2**s 
            b11=b10+2**t 
            if gate[0]=='cx':
                k[b10],k[b11]=k[b11],k[b10] 
            elif gate[0]=='crx':
                k[b10],k[b11]=turn(k[b10],k[b11],theta) 
            elif gate[0]=='swap':
                k[b01],k[b10]=k[b10],k[b01] 
  if get=='statevector':
    return k
  else:
    probs = [e[0]**2+e[1]**2 for e in k]
    if noise_model:
      for j in range(qc.num_qubits):
        p_meas = noise_model[j]
        for i0 in range(2**j):
          for i1 in range(2**(qc.num_qubits-j-1)):
            b0=i0+2**(j+1)*i1 
            b1=b0+2**j 
            p0 = probs[b0]
            p1 = probs[b1]
            probs[b0] = (1-p_meas)*p0 + p_meas*p1
            probs[b1] = (1-p_meas)*p1 + p_meas*p0
    if get=='probabilities_dict':
      return {('{0:0'+str(qc.num_qubits)+'b}').format(j):p for j,p in enumerate(probs)}
    elif get in ['counts', 'memory']:
      m = [False for _ in range(qc.num_qubits)]
      for gate in qc.data:
        for j in range(qc.num_qubits):
          assert  not ((gate[-1]==j) and m[j]), 'Incorrect or missing measure command.'
          m[j] = (gate==('m',j,j))
      m=[]
      for _ in range(shots):
        cumu=0
        un=True
        r=random.random()
        for j,p in enumerate(probs):
          cumu += p
          if r<cumu and un:    
            raw_out=('{0:0'+str(qc.num_qubits)+'b}').format(j)
            out_list = ['0']*qc.num_clbits
            for bit in outputnum_clbitsap:
              out_list[qc.num_clbits-1-bit] = raw_out[qc.num_qubits-1-outputnum_clbitsap[bit]]
            out = ''.join(out_list)
            m.append(out)
            un=False
      if get=='memory':
        return m
      else:
        counts = {}
        for out in m:
          if out in counts:
            counts[out] += 1
          else:
            counts[out] = 1
        return counts

File: versions/JavaScript/micromoth.js
// (C) Copyright Moth Quantum 2025.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

// (C) Copyright @kaleb-hutchy 2020.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

const r2 = Math.SQRT1_2; // 1/sqrt(2) for Hadamard gate

class QuantumCircuit {
    #data;
  
    constructor(n, m = 0) {
        if (!Number.isInteger(n) || n < 1 || n > 20) {
            throw new Error('num_qubits must be an integer between 1 and 20');
        }
        if (!Number.isInteger(m) || m < 0) {
            throw new Error('num_clbits must be a non-negative integer');
        }
        
        this.num_qubits = n;
        this.num_clbits = m;
        this.name = '';

        this.#data = [];
    };

    getData() {
        return this.#data;
    }

    initialize(state) (
        this.#data.push({ gate: "init", state });
        return this;
    }

    // Pauli-X gate
    x(q) {
        this.#data.push(['x', q]);
        return this;
    };

    // Hadamard gate
    h(q) {
        this.#data.push(['h', q]);
        return this;
    };

    // CX gate (Control-X gate)
    cx(s, t) { // source, target
        this.#data.push(['cx', s, t]);
        return this;
    };

    // Rotation-X gate
    rx(theta, q) {
        this.#data.push(['rx', theta, q]);
        return this;
    };

    // Rotation-Y gate
    ry(theta, q) {
        this.rx(Math.PI / 2, q);
        this.rz(theta, q);
        this.rx(-Math.PI / 2, q);
        return this;
    };

    // Rotation-Z gate
    rz(theta, q) {
        this.h(q);
        this.rx(theta, q);
        this.h(q);
        return this;
    };

    // Y gate
    y(q) {
        this.rz(Math.PI, q);
        this.x(q);
        return this;
    };

    // Pauli-Z gate
    z(q) {
        this.rz(Math.PI, q);
        return this;
    };

    // T gate
    t(q) {
        this.rz(Math.PI / 4, q);
        return this;
    };

    crx(theta, s, t) {
        this.#data.push(['crx', theta, s, t]);
        return this;
    };

    crz(theta, s, t) {
        this.#data.push(['crz', theta, s, t]);
        return this;
    }; 
    
    swap(s, t) {
        this.#data.push(['swap', s, t]);
        return this;
    };

    measure(q, b) {
        if (q >= this.num_qubits) {
            throw new Error(`Qubit index ${q} out of range. Must be < ${this.num_qubits}`);
        }

        if (b >= this.num_clbits) {
            throw new Error(`Classical bit index ${b} out of range. Must be < ${this.num_clbits}`);
        }

        this.#data.push(['m', q, b]);
        return this;
    };

    measure_all() {
        if (this.num_clbits === 0) this.num_clbits = this.num_qubits;
        for (let q = 0; q < this.num_qubits; q++) this.measure(q, q);

        return this;
    }
};

function simulate(qc, shots = 1024, get = 'counts', noise_model = []) {
    const nq = qc.num_qubits;
    const nc = qc.num_clbits;
    const d = qc.getData();

    // Limitation
    if (nq > 20) {
        console.warn(`Warning: Simulating ${nq} qubits requires ${Math.pow(2, nq).toLocaleString()} state amplitudes. This may cause performance issues or crash.`);
    }
    if (nq > 25) {
        throw new Error(`Cannot simulate ${nq} qubits (requires ${Math.pow(2, nq).toLocaleString()} state amplitudes). Maximum recommended: 20 qubits.`);
    }

    let state = Array(2 ** nq).fill(null).map(() => [0, 0]);
    state[0] = [1, 0]; // a |000...000> statevector

    if (!Array.isArray(noise_model)) {
        noise_model = Array(nq).fill(noise_model);
    }

    const outmap = {};

    for (let gate of d) {
        let op, args;

        if (Array.isArray(gate)) {
            [op, ...args] = gate;
        }
        else if (typeof gate == 'object' && gate.gate == 'init') {
            op = 'init';
            args = [gate.state];
        }
        else {
            throw new Error('Unknown gate format: ' + JSON.stringify(gate));
        }

        if (op == 'm') {
            outmap[args[1]] = args[0];
        }
        else if (op == 'init') {
            const s = args[0];

            for (let i = 0; i < s.length && i < state.length; i++) {
                state[i] = s[i];
            }
        }
        else if (['x', 'h', 'rx', 'rz'].includes(op)) {
            const j = args[args.length - 1];
            const mask = 1 << j;
            const step = 1 << (j + 1);
            
            // Nested loops to single iteration
            for (let i = 0; i < (1 << nq); i += step) {
                for (let k = 0; k < mask; k++) {
                    const b0 = i + k;
                    const b1 = b0 + mask;

                    if (op == 'x') {
                        [state[b0], state[b1]] = [state[b1], state[b0]];
                    }
                    else if (op == 'h') {
                        [state[b0], state[b1]] = superposition(state[b0], state[b1]);
                    }
                    else if (op == 'rx') {
                        [state[b0], state[b1]] = rotate(state[b0], state[b1], args[0]);
                    }
                    else if (op == 'rz') {
                        [state[b0], state[b1]] = phaseturn(state[b0], state[b1], args[0]);
                    }
                }
            }
        }
        else if (['cx', 'crx', 'swap'].includes(op)) {
            const theta = (op == 'crx') ? args[0] : null;
            const [s, t] = (op == 'crx') ? args.slice(1) : args;
            const [l, h] = [Math.min(s, t), Math.max(s, t)];
            
            // Cache bit shifts for performance
            const mask_l = 1 << l;
            const mask_s = 1 << s;
            const mask_t = 1 << t;
            const step_l = 1 << (l + 1);
            const step_h = 1 << (h + 1);

            // Iteration with cached masks
            for (let i = 0; i < (1 << nq); i += step_h) {
                for (let j = 0; j < (1 << h); j += step_l) {
                    for (let k = 0; k < mask_l; k++) {
                        const b00 = i + j + k;
                        const b01 = b00 + mask_t;
                        const b10 = b00 + mask_s;
                        const b11 = b10 + mask_t;

                        if (op == 'cx') {
                            [state[b10], state[b11]] = [state[b11], state[b10]];
                        }
                        else if (op == 'crx') {
                            [state[b10], state[b11]] = rotate(state[b10], state[b11], theta);
                        }
                        else if (op == 'swap') {
                            [state[b01], state[b10]] = [state[b10], state[b01]];
                        }
                    }
                }
            }
        }
    }

    if (get == 'statevector') return state;
    // ====================================

    let probs = state.map(([a, b]) => a * a + b * b);

    if (noise_model.length > 0) {
        for (let j = 0; j < nq; j++) {
            const p = noise_model[j];
            if (p === 0) continue; // Skip if no noise on this qubit
            
            const mask = 1 << j;
            const step = 1 << (j + 1);
            const p0_coeff = 1 - p;
            
            // Reduced nested loop with bit manipulation
            for (let i = 0; i < (1 << nq); i += step) {
                for (let k = 0; k < mask; k++) {
                    const b0 = i + k;
                    const b1 = b0 + mask;
                    const p0 = probs[b0], p1 = probs[b1];
                    probs[b0] = p0_coeff * p0 + p * p1;
                    probs[b1] = p0_coeff * p1 + p * p0;
                }
            }
         }
    }

    if (get == 'probability_dictionary') {
        return Object.fromEntries(
            probs.map((p, i) => [i.toString(2).padStart(nq, '0'), p])
        );
    }

    let results = [];
    const outIndices = Object.keys(outmap).map(Number);
    const hasMapping = outIndices.length > 0;

    for (let _ = 0; _ < shots; _++) {
        let r = Math.random();
        let cumu = 0;
        let chosen = 0;

        for (let j = 0; j < probs.length; j++) {
            cumu += probs[j];
            if (r < cumu) {
                chosen = j;
                break;
            }
        }

        if (hasMapping) {
            const raw = chosen.toString(2).padStart(nq, '0');
            let out_arr = Array(nc).fill('0');
            for (let bit of outIndices) {
                out_arr[nc - 1 - bit] = raw[nq - 1 - outmap[bit]];
            }
            results.push(out_arr.join(''));
        } else {
            results.push(chosen.toString(2).padStart(nc || nq, '0'));
        }
    }

    if (get == 'memory') return results; // accumulating changes

    const counts = {};
    for (let res of results) {
        counts[res] = (counts[res] || 0) + 1;
    }

    return counts;
};

function rotate(x, y, theta) {
    const COS = Math.cos(theta / 2);
    const SIN = Math.sin(theta / 2);

    return [
        [x[0] * COS + y[1] * SIN, x[1] * COS - y[0] * SIN],
        [y[0] * COS + x[1] * SIN, y[1] * COS - x[0] * SIN]
    ];
};

function superposition(x, y) {
    return [
        [r2 * (x[0] + y[0]), r2 * (x[1] + y[1])],
        [r2 * (x[0] - y[0]), r2 * (x[1] - y[1])]
    ]
};

function phaseturn(x, y, tt) {
    const COS = Math.cos(tt / 2);
    const SIN = Math.sin(tt / 2);

    return [
        [x[0] * COS + x[1] * SIN, x[1] * COS - x[0] * SIN],
        [y[0] * COS - y[1] * SIN, y[1] * COS + y[0] * SIN]
    ]
};

function kron(vec0, vec1) {
    const newVec = []
    for (const amp0 of vec0) {
        for (const amp1 of vec1) {
            newVec.push([amp0[0] * amp1[0] - amp0[1] * amp1[1], amp0[0] * amp1[1] + amp0[1] * amp1[0]]);    
        }
    }

    return newVec;
}

// Normalise a statevector
function norm(ket) {
    let n = 0
    for (const amp of ket) n += amp[0] * amp[0] + amp[1] * amp[1];
    n = Math.sqrt(n);

    return ket.map((amp) => [amp[0] / n, amp[1] / n]);
}

// Global exports for browser usage (p5.js compatible)
if (typeof window !== 'undefined') {
    window.QuantumCircuit = QuantumCircuit;
    window.simulate = simulate;
    window.rotate = rotate;
    window.superposition = superposition;
    window.phaseturn = phaseturn;
    window.kron = kron;
    window.norm = norm;
}